### PR TITLE
Initial draft

### PR DESCRIPTION
**Pull request by ian-bartholomew**
open for discussion and feedback

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,204 @@
...
+        - [Notes](#notes-1)
+        - [Objectives](#objectives)
+      - [Service](#service)
+
+## Introduction
+
+The intent of this document is to outline the OpenSLO specification.
+
+The goal of this project is to provide an open specification for defining and
+interfacing with SLOs to allow for a common, giving a set vendor-agnostic
```
Comment (niallrmurphy) [inline README.md:22]: s/for a common,/for a common approach,/
s/giving a set/and giving a/

File: README.md [hunk 2]
```diff
@@ -1 +1,204 @@
...
+solution to defining and tracking SLOs. Platform specific implimentation details
```
Comment (niallrmurphy) [inline README.md:23]: s/implimentation/implementation/

As a matter of interest, what _could_ be platform specific about this? Are you talking about the potential interactions with the monitoring system?

File: README.md [hunk 3]
```diff
@@ -1 +1,204 @@
...
+are purposefully excluded from the scope of this specification.
+
+## Specification
+
+### Goals
+
+- Compliance with the Kubernetes YAML format.
+- Vendor agnostic
```
Comment (niallrmurphy) [inline README.md:31]: - You dashed it above, so dash it here too (or undash below)
- I'm curious as to why k8s is a specific motivator here? Do you expect this to plug in directly?
- When you say "elsewhere", do you mean to enable forks from day zero? Surely we want the core system to be extensible (and extended) instead?

File: README.md [hunk 4]
```diff
@@ -1 +1,204 @@
...
+  - contain only lowercase alphanumeric characters or `-`
+  - start with an alphanumeric character
+  - end with an alphanumeric character
+
+---
+
+#### SLO
+
+A service level objective (SLO) is a target value or range of values for
+a service level that is measured by a service level indicator (SLI).
```
Comment (niallrmurphy) [inline README.md:66]: I wonder if "described" is a better word than "measured" here?

File: README.md [hunk 5]
```diff
@@ -1 +1,204 @@
...
+  name: string
+  displayName: string #optional
+  project: string
+spec:
+  description: string #optional
+  service: [service name] #name of the service to associate this SLO with
+  timeWindows:
+    # exactly one item, one of possible rolling time window or calendar aligned
+    # rolling time window
+    - unit: Day | Hour | Minute
```
Comment (niallrmurphy) [inline README.md:85]: I'm going to argue that the denomination should just be seconds here (and elsewhere). There's always weird corner-cases that don't fit into pre-determined boundaries.

File: README.md [hunk 6]
```diff
@@ -1 +1,204 @@
...
+Objectives are the thresholds for your SLOs. You can use objectives to define
+the tolerance levels for your metrics
+
+[TODO] This only covers ratio right now.  Since thresholds require the rawMetric,
+which is part of indicator stanza, we need to find a way of handling that.
+
+```yaml
+objectives:
+  - displayName: string # optional
+    value: numeric # value used to compare metrics values. All objectives of the SLO need to have unique value.
```
Comment (niallrmurphy) [inline README.md:153]: s/unique value/a unique value/ surely?

**Review by niallrmurphy — COMMENTED** [review]
Initial comments on spec.

File: README.md (hunk 2 above)
Comment (ian-bartholomew) [reply README.md:23]: > As a matter of interest, what could be platform specific about this? Are you talking about the potential interactions with the monitoring system?

Yeah, exactly those sorts of things.  Like for us, we have agent configuration, and that might look different depending on the vendor

File: README.md (hunk 3 above)
Comment (ian-bartholomew) [reply README.md:31]: > I'm curious as to why k8s is a specific motivator here? Do you expect this to plug in directly?

 I think that for us using that format checked a lot of boxes.  Moving forward, I don't imagine that we we would have it plug in directly, no

File: README.md (hunk 5 above)
Comment (ian-bartholomew) [reply README.md:85]: I agree.  Another future improvement might be to later add support for abbreviated units, like `1h` or `24s`

File: README.md (hunk 1 above)
Comment (ian-bartholomew) [reply README.md:22]: fixed

File: README.md (hunk 2 above)
Comment (ian-bartholomew) [reply README.md:23]: fixed `implementation`

File: README.md (hunk 3 above)
Comment (ian-bartholomew) [reply README.md:31]: fixed

File: README.md (hunk 6 above)
Comment (ian-bartholomew) [reply README.md:153]: fixed

File: README.md (hunk 3 above)
Comment (kitmerker) [reply README.md]: > I'm curious as to why k8s is a specific motivator here? Do you expect this to plug in directly?

I would say the goal is to be idiomatic of Kubernetes design, for the same reason @ian-bartholomew describes; a lot of thought/consensus has gone into it, so we can benefit from those design decisions and make the format familiar in structure. We should not constrain ourselves to k8s yaml but take it as inspiration.

**Pull request merged**